File: 239. Sliding Window Maximum/c++_20220902/tO(n)_sO(k).cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int n = nums.size();
        deque<int> pq; //[(index)]
        vector<int> ans;
        for(int i=0;i<n;++i){
            while((pq.empty()==false) && (nums[i] >= nums[pq.back()])){
                pq.pop_back();
            }
            pq.push_back(i);
            if(i>=k-1){
                while(pq.front() <= i-k){
                    pq.pop_front();
                }
                ans.push_back(nums.at(pq.front()));
            }
        }
        return ans;
    }
};

/*
Runtime: 328 ms, faster than 88.75% of C++ online submissions for Sliding Window Maximum.
Memory Usage: 134.6 MB, less than 92.32% of C++ online submissions for Sliding Window Maximum.
*/

File: 239. Sliding Window Maximum/python_20220902/tO(n)_sO(k).py
class Solution(object):
    def maxSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        n = len(nums)
        q = collections.deque()
        for i in range(k):
            while q and nums[i] >= nums[q[-1]]:
                # print(nums[i], nums[q[-1]])
                q.pop()
            q.append(i)
        ans = [nums[q[0]]]
        for i in range(k, n):
            while q and nums[i] >= nums[q[-1]]:
                q.pop()
            q.append(i)
            while q[0] <= i-k:
                q.popleft()
            ans.append(nums[q[0]])
        return ans